More aggressive runtime code exclusion.

chriseth opened the pull request:
Fixes https://github.com/ethereum/solidity/issues/13680

chriseth committed:
More aggressive runtime code exclusion.

chriseth committed:
Gas updates.

cameel commented on test/libsolidity/semanticTests/userDefinedValueType/erc20.sol:
```diff
@@ -114,8 +114,8 @@ contract ERC20 {
 // gas irOptimized: 362054
-// gas legacy: 852973
-// gas legacyOptimized: 419237
+// gas legacy: 852825
+// gas legacyOptimized: 418837
```
@NunoFilipeSantos @franzihei What do you think about including the cost and bytecode size comparison table in the release post? We can easily get one from our benchmarks on external test and it would look like the one in the [0.8.15 announcement](https://blog.soliditylang.org/2022/06/15/solidity-0.8.15-release-announcement/), except showing very slight differences. It might put people at ease regarding the severity of this. The gas differences here seem minimal (way below 1%) and it's likely the same in real-life projects.

franzihei replied:
Yes I think that's a great idea.

cameel replied:
ok then. We need to get ext tests fixed on `develop` to get all benchmarks so we should wait with merging this until #13975 is fixed.

NunoFilipeSantos replied:
Being technically honest in blogposts is great for any dev tools.

NunoFilipeSantos replied:
Even if it means sometimes we say that we lag behind some other alternative.

cameel commented on test/libsolidity/semanticTests/various/runtime_code_exclusion.sol:
```diff
@@ -0,0 +1,34 @@
+contract C {
+    bytes data;
+    constructor() {
+        data = L.longdata();
```
There are several different paths through `ContractCompiler` that now make a decision about including the label. This tests only one of them (member access for non-attached functions). Can you add cases covering the remaining ones? I.e. attached functions, member access on a module and non-member-access.

chriseth replied:
Note that we are just using the same mechanism we are using as well in the ir-based compilation.

cameel replied:
I'm not doubting it works, I just want all important paths covered because this looks like something that someone can easily break in the codegen and not notice because it would not lead to an error.

cameel commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1541,7 +1523,11 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
+								*function,
+								// If we call directly, do not include the second label.
+								!_memberAccess.annotation().calledDirectly
+							);
```
There's code handling a `super` as a special case above. It uses `pushCombinedFunctionEntryLabel()` as well. Doesn't it need to be adjusted?

Even if it doesn't it would be good to have a test case covering it.

chriseth replied:
I thought it to be too edge to change it, but I can take another look.

cameel reviewed (commented):
Looks good in terms of fixing it. I have some extra test suggestions and also a question about a potential corner case.

chriseth commented:
Added tests. For each call to 'pushCombinedFunctionEntryLabel' I changed the code individually to unconditionally return "true", and each time one of the tests failed.

cameel commented on test/libsolidity/semanticTests/deployedCodeExclusion/static_base_function_deployed.sol:
```diff
@@ -0,0 +1,31 @@
+    bytes data;
+    function () pure returns (bytes memory) f;
+    constructor() {
+        data = super.longdata();
+        f = super.longdata;
```
Looks like this was meant to be in `super_function_deployed.sol` and the corresponding part in `static_base_function_deployed.sol` should be here instead :)

chriseth replied:
Ah nice! Fixed now.

cameel reviewed (commented):
Just a small correction needed.

We still have a moment until the release though so I'd like us to fix #13975 on `develop` first so that we can get proper benchmarks before and after (currently OZ ext test is failing and it's an important one).

matheusaaguiar commented:
Rebased.

matheusaaguiar commented:
Benchmark diff 
### `ir-no-optimize`
|          project | bytecode_size | deployment_gas | method_gas |
|:----------------:|--------------:|---------------:|-----------:|
|           bleeps |               |                |            |
|            brink |               |                |            |
|        chainlink |               |                |            |
|           colony |          `0%` |                |            |
|        elementfi |               |                |            |
|              ens |               |                |            |
|            euler |               |                |            |
|           gnosis |               |                |            |
|              gp2 |               |                |            |
|  perpetual-pools |               |                |            |
|    pool-together |               |                |            |
|         prb-math |               |                |            |
|          trident |               |                |            |
|          uniswap |               |                |            |
| yield_liquidator |               |                |            |
|         zeppelin |               |                |            |

### `ir-optimize-evm+yul`
|          project | bytecode_size | deployment_gas | method_gas |
|:----------------:|--------------:|---------------:|-----------:|
|           bleeps |          `0%` |           `0%` |       `0%` |
|            brink |          `0%` |                |            |
|        chainlink |          `0%` |           `0%` |      `+0%` |
|           colony |          `0%` |                |            |
|        elementfi |          `0%` |                |            |
|              ens |          `0%` |           `0%` |       `0%` |
|            euler |          `0%` |          `-0%` |      `-0%` |
|           gnosis |          `0%` |           `0%` |      `+0%` |
|              gp2 |          `0%` |           `0%` |      `+0%` |
|  perpetual-pools |          `0%` |          `-0%` |      `-0%` |
|    pool-together |          `0%` |          `-0%` |      `-0%` |
|         prb-math |          `0%` |          `+0%` |       `0%` |
|          trident |          `0%` |          `-0%` |      `+0%` |
|          uniswap |          `0%` |          `+0%` |      `-0%` |
| yield_liquidator |          `0%` |          `+0%` |       `0%` |
|         zeppelin |          `0%` |          `-0%` |      `-0%` |

### `ir-optimize-evm-only`
|          project | bytecode_size | deployment_gas | method_gas |
|:----------------:|--------------:|---------------:|-----------:|
|           bleeps |               |                |            |
|            brink |               |                |            |
|        chainlink |               |                |            |
|           colony |          `0%` |                |            |
|        elementfi |               |                |            |
|              ens |               |                |            |
|            euler |               |                |            |
|           gnosis |               |                |            |
|              gp2 |               |                |            |
|  perpetual-pools |               |                |            |
|    pool-together |               |                |            |
|         prb-math |               |                |            |
|          trident |               |                |            |
|          uniswap |               |                |            |
| yield_liquidator |               |                |            |
|         zeppelin |               |                |            |

### `legacy-no-optimize`
|          project |  bytecode_size | deployment_gas |     method_gas |
|:----------------:|---------------:|---------------:|---------------:|
|           bleeps |                |                |                |
|            brink |           `0%` |                |                |
|        chainlink | **`-0.01% ✅`** |                |                |
|           colony |           `0%` |                |                |
|        elementfi | **`-1.16% ✅`** |                |                |
|              ens |           `0%` |                |                |
|            euler | **`-0.08% ✅`** | **`-0.09% ✅`** |          `-0%` |
|           gnosis |           `0%` |          `+0%` |          `+0%` |
|              gp2 | **`-0.03% ✅`** |                |                |
|  perpetual-pools |           `0%` |          `+0%` | **`+0.01% ❌`** |
|    pool-together |           `0%` |          `-0%` |          `-0%` |
|         prb-math |           `0%` |          `-0%` |           `0%` |
|          trident | **`-0.15% ✅`** |          `+0%` |          `-0%` |
|          uniswap | **`-0.43% ✅`** | **`-0.41% ✅`** | **`-0.66% ✅`** |
| yield_liquidator |           `0%` |           `0%` |           `0%` |
|         zeppelin | **`-1.56% ✅`** | **`-1.81% ✅`** |          `+0%` |

### `legacy-optimize-evm+yul`
|          project |  bytecode_size | deployment_gas |     method_gas |
|:----------------:|---------------:|---------------:|---------------:|
|           bleeps |           `0%` |           `0%` |           `0%` |
|            brink |           `0%` |                |                |
|        chainlink | **`-0.02% ✅`** | **`-0.01% ✅`** |          `-0%` |
|           colony |           `0%` |                |                |
|        elementfi | **`-1.18% ✅`** |                |                |
|              ens |           `0%` | **`-0.06% ✅`** | **`-0.02% ✅`** |
|            euler | **`-0.03% ✅`** | **`-0.03% ✅`** |          `-0%` |
|           gnosis |           `0%` |           `0%` |          `+0%` |
|              gp2 | **`-0.07% ✅`** | **`-0.04% ✅`** |          `+0%` |
|  perpetual-pools |           `0%` |          `-0%` |          `-0%` |
|    pool-together |           `0%` |          `-0%` |          `+0%` |
|         prb-math |           `0%` |          `-0%` |           `0%` |
|          trident | **`-0.13% ✅`** |          `+0%` |          `-0%` |
|          uniswap | **`-0.47% ✅`** | **`-0.44% ✅`** | **`-0.64% ✅`** |
| yield_liquidator |           `0%` |          `+0%` |           `0%` |
|         zeppelin | **`-1.65% ✅`** | **`-1.47% ✅`** |          `-0%` |

### `legacy-optimize-evm-only`
|          project |  bytecode_size | deployment_gas |     method_gas |
|:----------------:|---------------:|---------------:|---------------:|
|           bleeps |                |                |                |
|            brink |           `0%` |                |                |
|        chainlink | **`-0.02% ✅`** | **`-0.01% ✅`** |          `-0%` |
|           colony |           `0%` |                |                |
|        elementfi | **`-1.07% ✅`** |                |                |
|              ens |           `0%` | **`-0.06% ✅`** | **`-0.02% ✅`** |
|            euler | **`-0.02% ✅`** | **`-0.02% ✅`** |          `-0%` |
|           gnosis |           `0%` |          `+0%` |          `-0%` |
|              gp2 | **`-0.06% ✅`** | **`-0.04% ✅`** |          `+0%` |
|  perpetual-pools |           `0%` |          `+0%` |          `+0%` |
|    pool-together |           `0%` |          `-0%` |          `+0%` |
|         prb-math |           `0%` |          `+0%` |           `0%` |
|          trident | **`-0.11% ✅`** |          `-0%` |          `+0%` |
|          uniswap | **`-0.43% ✅`** |  **`-0.4% ✅`** | **`-0.61% ✅`** |
| yield_liquidator |           `0%` |          `-0%` |           `0%` |
|         zeppelin | **`-1.72% ✅`** | **`-1.68% ✅`** |          `-0%` |


`!V` = version mismatch
`!B` = no value in the "before" version
`!A` = no value in the "after" version
`!T` = one or both values were not numeric and could not be compared
`-0` = very small negative value rounded to zero
`+0` = very small positive value rounded to zero
